### PR TITLE
Run verification code before shape inference

### PR DESCRIPTION
**Pull request by etiotto**
Currently `ShapeInferencePass` calls `inferShapes` for each operation in a MLIR Region. This causes `inferShapes` to be run on an operation **after** has been run on all of its operands. It is assumed that the operation is legal once its `inferShapes` function is run, however this is not the case for operations that had operands with unknown shape and rank when the verification code was first executed (the verification code was skipped in that situation). 

This PR causes the verification code to be called on an operation immediately **before** its `inferShape` function is called. This guarantees that the proper diagnostics are emitted for operations that had operands with unknown shape when the verification code was first executed.

### CODE REVIEW

**Review by AlexandreEichenberger — COMMENTED** [review]
@chentong319 @etiotto My understanding was that a verifier is automatically called when we create any ops? Not opposed to add passes, just trying to understand

**Comment by etiotto** [comment]
> @chentong319 @etiotto My understanding was that a verifier is automatically called when we create any ops? Not opposed to add passes, just trying to understand

ShapeInferencePass let me try to clarify. This PR doesn't add a new pass. During the existing `ShapeInferencePass` it calls the verification code for an operator before shape inference is attempted for that operator. At the program point where I have added the call to the verification code the shape of all operands should be known because the algorithm runs shape inference on the operands that feed an operator before running it on the operator itself.

File: src/Transform/ONNX/ShapeInferencePass.cpp [hunk 1]
```diff
@@ -93,34 +97,45 @@ class ShapeInferencePass : public mlir::PassWrapper<ShapeInferencePass,
...
+        continue;
+
+      if (auto shape_op = llvm::dyn_cast<ShapeInference>(op)) {
+        LLVM_DEBUG(
+            llvm::dbgs() << DEBUG_TYPE << ": op " << op.getName() << "\n");
+        // Verify the operation before attempting to infer the shape of the
+        // produced output(s).
+        Optional<RegisteredOperationName> registeredInfo =
+            op.getName().getRegisteredInfo();
+        if (registeredInfo && failed(registeredInfo->verifyInvariants(&op)))
```
Comment (tungld) [inline src/Transform/ONNX/ShapeInferencePass.cpp:122]: Does this `verifyInvariants` finally call our custom `verify` function? Or do we need to override it for each op?

Comment (etiotto) [reply src/Transform/ONNX/ShapeInferencePass.cpp:122]: It does call the custom verify member function:
```
  static LogicalResult verifyInvariants(Operation *op) {
    static_assert(hasNoDataMembers(),
                  "Op class shouldn't define new data members");
    return failure(
        failed(op_definition_impl::verifyTraits<VerifiableTraitsTupleT>(op)) ||
        failed(cast<ConcreteType>(op).verify()));   <<<<<< HERE
  }
  ```

Comment (tungld) [reply src/Transform/ONNX/ShapeInferencePass.cpp:122]: Thanks for pointing it out!

**Comment by tungld** [comment]
> When the verification code was called in the first place, not all input operands might have a know shape/rank and so not all constraints could be verified. With this change the verification code is called again, when the input operands have known shape/rank to allow diagnosing constraints that we could not diagnose in the first call.

Are you sure that ShapeInferencePass will call verification code of an operation multiple times. It seems that ShapeInferencePass visits operations once in order.

**Comment by AlexandreEichenberger** [comment]
> Are you sure that ShapeInferencePass will call verification code of an operation multiple times. It seems that ShapeInferencePass visits operations once in order.

@chentong319 can you chime in?

**Comment by chentong319** [comment]
The verifier is called automatically after each pass.

**Comment by AlexandreEichenberger** [comment]
Personally, I am still confused.
* if it is run before shape inference, then its likely the rank is known but not the sizes of each dimension
* after shape inference, the rank and size should be known if not runtime dims.
So how does running verify before shape inference help, I don't know.

In addition, I though that one the bug we were trying to fix initially was when the rank was ridiculously large (because of an endianness issue).  This can be done regardless of whether the dimensions are known (though we say >100 or 1k, flag it at least with a warning).

**Comment by etiotto** [comment]
> Personally, I am still confused.
> * if it is run before shape inference, then its likely the rank is known but not the sizes of each dimension
> * after shape inference, the rank and size should be known if not runtime dims.
So how does running verify before shape inference help, I don't know.

This is how the `onnx.Concat operation looks like before `DecomposeONNXToONNXPass`, the rank of the operands is unknown, we cannot diagnose that `axis` is in range [-r, r-1].
```
  %76 = "onnx.Concat"(%66, %75, %22, %24, %39, %41, %15, %17, %31, %33, %6, %57, %55, %53, %51, %49, %47)
   {axis = 33554432 : si64, onnx_node_name = "StatefulPartitionedCall/model_4/tf.concat_1/concat"} 
   : (tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>, tensor<*xf32>) -> tensor<*xf32>
```

After we run shape inference for the operations that feed the Concat, we determine the rank of all the input operands. 
At that point, before running shape inference on `onnx.Concat`, we can call verify() on Concat, because now all ranks of the inputs are known, the verify() code can diagnose precisely that the rank is out of bound. 

> In addition, I though that one the bug we were trying to fix initially was when the rank was ridiculously large (because of > an endianness issue).  This can be done regardless of whether the dimensions are known (though we say >100 or 1k, 
> flag it at least with a warning).

It can be done but this approach has the advantage of being able to report the exact range `axis` should belong to. 
This PR gives us a way to diagnose error conditions that require knowing the shape of the inputs, if it isn't know to start with.

**Comment by chentong319** [comment]
I got the point why verifier is needed explicitly at the beginning of shapeInference function for some ops. Will the error be caught at the end of the shape inference pass if the verifier is not added?

**Comment by chentong319** [comment]
For implementation, can we just call verifier before every call of shape inference?
In general, I prefer a simple fix if possible.  Avoid unnecessary changes, which will put extra burden on reviewers.

**Comment by AlexandreEichenberger** [comment]
@adriansion 

I ran an example of your test like this:
``` shell
python ~/onnx-mlir/test/perf/ParseBenchmarks.py  --compare  PerfGemm_Benchmark_1650564009 PerfGemm_Benchmark_1650564009 Pct --verbose
```
and got this 
``` shell
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare/1024"
39                     4
18.5847                149.555
18.5738                149.509
ms                     ms
2.68435e+08            2.14748e+09
1.44524e+10            1.43635e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare/2048"
39                     1
18.5847                4262.96
18.5738                4260.45
ms                     ms
2.68435e+08            1.71799e+10
1.44524e+10            4.03241e+09
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare_BigO"
18.5847                0.490696
18.5738                0.490407
ms                     N^3
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare_RMS"
18.5847                0.244589
18.5738                0.244559
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/16"
39                     1225340
18.5847                0.000611784
18.5738                0.0006112
ms                     ms
2.68435e+08            7936
1.44524e+10            1.29843e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/32"
39                     256126
18.5847                0.00276785
18.5738                0.00276213
ms                     ms
2.68435e+08            64512
1.44524e+10            2.33559e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/64"
39                     29550
18.5847                0.0219624
18.5738                0.021894
ms                     ms
2.68435e+08            520192
1.44524e+10            2.37595e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/128"
39                     4619
18.5847                0.149305
18.5738                0.149138
ms                     ms
2.68435e+08            4.17792e+06
1.44524e+10            2.80138e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/256"
39                     469
18.5847                1.40377
18.5738                1.39973
ms                     ms
2.68435e+08            3.34889e+07
1.44524e+10            2.39252e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/512"
39                     58
18.5847                10.9099
18.5738                10.8869
ms                     ms
2.68435e+08            2.68173e+08
1.44524e+10            2.46326e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/1024"
39                     7
18.5847                94.0158
18.5738                93.8342
ms                     ms
2.68435e+08            2.14644e+09
1.44524e+10            2.28748e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/2048"
39                     1
18.5847                749.031
18.5738                746.102
ms                     ms
2.68435e+08            1.71757e+10
1.44524e+10            2.30206e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare_BigO"
18.5847                0.0872028
18.5738                0.0868644
ms                     N^3
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare_RMS"
18.5847                0.00292568
18.5738                0.00318598
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/16"
39                     1006534
18.5847                0.00067036
18.5738                0.000670227
ms                     ms
2.68435e+08            8192
1.44524e+10            1.22227e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/32"
39                     214898
18.5847                0.00313916
18.5738                0.00313888
ms                     ms
2.68435e+08            65536
1.44524e+10            2.08788e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/64"
39                     29130
18.5847                0.022712
18.5738                0.0226865
ms                     ms
2.68435e+08            524288
1.44524e+10            2.31101e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/128"
39                     4387
18.5847                0.155923
18.5738                0.155919
ms                     ms
2.68435e+08            4.1943e+06
1.44524e+10            2.69005e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/256"
39                     474
18.5847                1.41001
18.5738                1.40963
ms                     ms
2.68435e+08            3.35544e+07
1.44524e+10            2.38037e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/512"
39                     58
18.5847                12.1311
18.5738                12.1112
ms                     ms
2.68435e+08            2.68435e+08
1.44524e+10            2.21643e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/1024"
39                     6
18.5847                94.4452
18.5738                94.2716
ms                     ms
2.68435e+08            2.14748e+09
1.44524e+10            2.27797e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/2048"
39                     1
18.5847                772.221
18.5738                771.601
ms                     ms
2.68435e+08            1.71799e+10
1.44524e+10            2.22652e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare_BigO"
18.5847                0.0898686
18.5738                0.089795
ms                     N^3
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare_RMS"
18.5847                0.0066506
18.5738                0.00696144
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare/1024"
39                     4
18.5847                149.555
18.5738                149.509
ms                     ms
2.68435e+08            2.14748e+09
1.44524e+10            1.43635e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare/2048"
39                     1
18.5847                4262.96
18.5738                4260.45
ms                     ms
2.68435e+08            1.71799e+10
1.44524e+10            4.03241e+09
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare_BigO"
18.5847                0.490696
18.5738                0.490407
ms                     N^3
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatmulSquare_RMS"
18.5847                0.244589
18.5738                0.244559
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/16"
39                     1225340
18.5847                0.000611784
18.5738                0.0006112
ms                     ms
2.68435e+08            7936
1.44524e+10            1.29843e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/32"
39                     256126
18.5847                0.00276785
18.5738                0.00276213
ms                     ms
2.68435e+08            64512
1.44524e+10            2.33559e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/64"
39                     29550
18.5847                0.0219624
18.5738                0.021894
ms                     ms
2.68435e+08            520192
1.44524e+10            2.37595e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/128"
39                     4619
18.5847                0.149305
18.5738                0.149138
ms                     ms
2.68435e+08            4.17792e+06
1.44524e+10            2.80138e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/256"
39                     469
18.5847                1.40377
18.5738                1.39973
ms                     ms
2.68435e+08            3.34889e+07
1.44524e+10            2.39252e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/512"
39                     58
18.5847                10.9099
18.5738                10.8869
ms                     ms
2.68435e+08            2.68173e+08
1.44524e+10            2.46326e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/1024"
39                     7
18.5847                94.0158
18.5738                93.8342
ms                     ms
2.68435e+08            2.14644e+09
1.44524e+10            2.28748e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare/2048"
39                     1
18.5847                749.031
18.5738                746.102
ms                     ms
2.68435e+08            1.71757e+10
1.44524e+10            2.30206e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare_BigO"
18.5847                0.0872028
18.5738                0.0868644
ms                     N^3
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_MatMulWithGemmSquare_RMS"
18.5847                0.00292568
18.5738                0.00318598
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/16"
39                     1006534
18.5847                0.00067036
18.5738                0.000670227
ms                     ms
2.68435e+08            8192
1.44524e+10            1.22227e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/32"
39                     214898
18.5847                0.00313916
18.5738                0.00313888
ms                     ms
2.68435e+08            65536
1.44524e+10            2.08788e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/64"
39                     29130
18.5847                0.022712
18.5738                0.0226865
ms                     ms
2.68435e+08            524288
1.44524e+10            2.31101e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/128"
39                     4387
18.5847                0.155923
18.5738                0.155919
ms                     ms
2.68435e+08            4.1943e+06
1.44524e+10            2.69005e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/256"
39                     474
18.5847                1.41001
18.5738                1.40963
ms                     ms
2.68435e+08            3.35544e+07
1.44524e+10            2.38037e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/512"
39                     58
18.5847                12.1311
18.5738                12.1112
ms                     ms
2.68435e+08            2.68435e+08
1.44524e+10            2.21643e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/1024"
39                     6
18.5847                94.4452
18.5738                94.2716
ms                     ms
2.68435e+08            2.14748e+09
1.44524e+10            2.27797e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare/2048"
39                     1
18.5847                772.221
18.5738                771.601
ms                     ms
2.68435e+08            1.71799e+10
1.44524e+10            2.22652e+10
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare_BigO"
18.5847                0.0898686
18.5738                0.089795
ms                     N^3
----------------------------------------
# arg = PerfGemm_Benchmark_1650564009
"BM_MatmulSquare/512"  "BM_GemmSquare_RMS"
18.5847                0.0066506
18.5738                0.00696144
----------------------------------------
Traceback (most recent call last):
  File "/workdir/onnx-mlir/test/perf/ParseBenchmarks.py", line 480, in <module>
    main()
  File "/workdir/onnx-mlir/test/perf/ParseBenchmarks.py", line 472, in main
    ComparisonOutput = CompareOutput(RawBenchmarkOutput1, RawBenchmarkOutput2, metric, maxrelativeslowdown)
  File "/workdir/onnx-mlir/test/perf/ParseBenchmarks.py", line 305, in CompareOutput
    dict1name = OutputDict1["name"]
KeyError: 'name'
root@1250b049ac2b:~/onnx-mlir/build/test/perf# python ~/onnx-mlir/test/perf/ParseBenchmarks.py  --compare  PerfGemm_Benchmark_1650564009 PerfGemm_Benchmark_1650564009 Pct --verbose
```

Any idea what's up?

**Comment by etiotto** [comment]
> I got the point why verifier is needed explicitly at the beginning of shapeInference function for some ops. Will the error be caught at the end of the shape inference pass if the verifier is not added?

If we do not call the verifier for the operation and we just call shape inference (like we do now) then during shape inference we trip into an assert, this happens  because the assumption the shape inference code makes (at least for the Concat operation, but likely for others as well) is that the operations is valid.

**Comment by etiotto** [comment]
> For implementation, can we just call verifier before every call of shape inference? In general, I prefer a simple fix if possible. Avoid unnecessary changes, which will put extra burden on reviewers.

Yes. This proposed change does exactly that.

**Comment by etiotto** [comment]
> > After we run shape inference for the operations that feed the Concat, we determine the rank of all the input operands.
> 
> @etiotto Oh, so shape inference run once on all nodes, and if there was one or more change it runs again. What you are suggesting is that before running it again, we insert a verify pass? Please confirm, then I fully understand the proposed change.

@AlexandreEichenberger  that is what I suggest, so that we have a change to diagnose a potential semantic violation. 
With this approach we don't have to worry about emitting diagnostic in the shape inference code, we can just reuse the diagnostic checks implemented in the `verify()` member function of an operator (if one exist).  I think that is a plus, because we can centralize the responsibility for diagnosing error conditions in one place.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, please @tungld or @chentong319 chime in if you still have reservations

**Comment by chentong319** [comment]
1. For historical reason, some of the semantic check is in shape inference. Assume that those checks are all moved to verifier. Can shape inference code run through and check the semantics later? This approach is risky, I know. If the error will be caught in shape inference, is it a sign of redundant code?
2. For implementation, the verifier is not called directly, but through an interface. Do we really need that interface?

**Comment by etiotto** [comment]
> For historical reason, some of the semantic check is in shape inference. Assume that those checks are all moved to verifier. Can shape inference code run through and check the semantics later? This approach is risky, I know. If the error will be caught in shape inference, is it a sign of redundant code?

The current (historical) approach of diagnosing semantic violations in the shape inference code is brittle IMO. Is not immediately clear where to add the checks within the shape inference code.  Also it leads to code duplication.  Placing all diagnostic code in the verify() member function of an operator prehempt code duplication, allow a reader to easily determine whether all constraints for an operation have been checked, and allow the same error condition to be diagnosed with a consistent message. 

> For implementation, the verifier is not called directly, but through an interface. Do we really need that interface?

**Comment by etiotto** [comment]
@chentong319 @tungld  I see Alex approved the PR, are all your comments been addressed to your satisfaction?

**Review by chentong319 — APPROVED** [review]
Thanks for explanation.

**Comment by etiotto** [comment]
> Thanks for explanation.

Thank you @chentong319 , @AlexandreEichenberger and @tungld for the through review and helpful comments.

**Review by tungld — APPROVED** [review]
LGTM. Thanks!

**Comment by etiotto** [comment]
@jenkins-droid please test this

**Comment by etiotto** [comment]
@jenkins-droid  test this please.

**Comment by etiotto** [comment]
@jenkins-droid retest this please.

**Comment by etiotto** [comment]
Currently ShapeInferencePass calls inferShapes for each operation in a MLIR Region. This causes inferShapes to be run on an operation after has been run on all of its operands. It is assumed that the operation is legal once its inferShapes function is run, however this is not the case for operations that had operands with unknown shape and rank when the verification code was first executed (the verification code was skipped in that situation).

This PR causes the verification code to be called on an operation immediately before its inferShape function is called. This guarantees that the proper diagnostics are emitted for operations that had operands with unknown shape when the verification code was first executed.

**Pull request merged**